presto-raptorx/src/main/java/com/facebook/presto/raptorx/storage/ChunkRecoveryManager.java

```diff
@@ -0,0 +1,427 @@
+import static java.nio.file.StandardCopyOption.ATOMIC_MOVE;
+import static java.util.Comparator.comparingInt;
+import static java.util.Objects.requireNonNull;
+import static java.util.UUID.randomUUID;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static java.util.stream.Collectors.toSet;
+
+public class ChunkRecoveryManager
+{
+    private static final Logger log = Logger.get(ChunkRecoveryManager.class);
+
+    private final StorageService storageService;
+    private final ChunkStore chunkStore;
+    private final String nodeIdentifier;
+    private final ChunkManager chunkManager;
+    private final Duration missingChunkDiscoveryInterval;
+
+    private final AtomicBoolean started = new AtomicBoolean();
+    private final MissingChunksQueue chunkQueue;
+
+    private final ScheduledExecutorService missingChunkExecutor = newSingleThreadScheduledExecutor(daemonThreadsNamed("missing-chunk-discovery"));
+    private final ExecutorService executorService = newCachedThreadPool(daemonThreadsNamed("chunk-recovery-%s"));
+    private final ChunkRecoveryStats stats;
+
+    @Inject
+    public ChunkRecoveryManager(
+            StorageService storageService,
+            ChunkStore chunkStore,
+            NodeManager nodeManager,
+            ChunkManager chunkManager,
+            ChunkRecoveryConfig config)
+    {
+        this(storageService,
+                chunkStore,
+                nodeManager,
+                chunkManager,
+                config.getMissingChunkDiscoveryInterval(),
+                config.getRecoveryThreads());
+    }
+
+    public ChunkRecoveryManager(
+            StorageService storageService,
+            ChunkStore chunkStore,
+            NodeManager nodeManager,
+            ChunkManager chunkManager,
+            Duration missingChunkDiscoveryInterval,
+            int recoveryThreads)
+    {
+        this.storageService = requireNonNull(storageService, "storageService is null");
+        this.chunkStore = requireNonNull(chunkStore, "chunkStore is null");
+        this.nodeIdentifier = requireNonNull(nodeManager, "nodeManager is null").getCurrentNode().getNodeIdentifier();
+        this.chunkManager = requireNonNull(chunkManager, "chunkManager is null");
+        this.missingChunkDiscoveryInterval = requireNonNull(missingChunkDiscoveryInterval, "missingChunkDiscoveryInterval is null");
+        this.chunkQueue = new MissingChunksQueue(new PrioritizedFifoExecutor<>(executorService, recoveryThreads, RecoveryRunnable.comparator()));
+        this.stats = new ChunkRecoveryStats();
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (!started.getAndSet(true)) {
+            scheduleRecoverMissingChunks();
+        }
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        executorService.shutdownNow();
```

Comment:
Given that `executorService` is actually being used two levels deeper (inside `PrioritizedFifoExecutor `, which is inside `MissingChunksQueue `), and nowhere else in this class, I feel like this should be pushed down to these classes.

Reply:
You don't want utility classes to create their own executors (for several reasons ... lifecycle, you might want to share them, what to name threads, etc.), so it needs to be created here.